### ISSUE TITLE
Module documentation

### ISSUE THREAD

**Issue by superihippo** [superihippo/RPFramework#2]
# RPFramework
Modular roleplaying mission framework for Arma 3. For license, check the LICENSE.txt file.

## Benefits of a modular approach
So why should you hop on board with the modular approach to this framework? 
This modular approach isn't a good approach for a normal mission but for a framework it has been proven effective. Think of Arma 3 as a framework and mods as modules; modules can be easily enabled and disabled by the user and, of course, installed easily without messing with the core files of the game. If you got the example you might already think it's a good approach for this framework as proven by Arma. In the RPFramework installing modules isn't as easy as in Arma 3, but it gets pretty close. I have seen the attempts of modifying missions like Altis Life that people have tried to use as a framework and it gets really messy, but this way you keep your modules and original files from the core separated.

### General
Modules are additional script packages that might depend on the core modules, but the core never depends on the module. Modules can overwrite settings, add functionality, add interaction menu items/options and much more. Modules can be distributed under different licenses than the core of RPFramework.
The core is a bunch of basic functionality that creates a fairly limited game experience and loads in modules.

Modules are divided into "Default Modules" and "Additional Modules". Default modules add default functionality to the framework, but can be easily disabled if one wants to change their functionality by an additional module. The difference between core and Default Modules is that core is standalone, but Default Modules depend on core and default modules can, of course, be disabled and it won't break the core.

### Installing modules
Installing additional modules should be instructed by the module creator as different modules might have different features but there is a general instruction:
1. Move module's files to mission>Functions>modules and RPF_Server>server>modules
2. Add module's function class into mission>Functions.hpp in the class ClientModules and into RPF_Server>config.cpp in the class ServerModules
3. Open up mission>Functions>core>init>fn_initModules.sqf and RPF_server>server>core>init>fn_initModules.sqf and in the mission's file add your module's init function to the additional modules array in format "ClientModule_fnc_initNameOfYourModule" and in the server's file add your module's init function to the additional modules array in format "ServerModule_fnc_initNameOfYourModule".
4. Your module should be installed properly.

### Creating a module
Modules add features to the RPFramework. If you are a scripter, you already know how to make modules because they are just bunch of functions to the mission and server.

Modules aren't anything special, but they just make things a bit more modular and easier for the user of the framework.

You can add any functions, sounds and dialogs if you just make sure to instruct the user of your module to add them properly. It is also very important that you follow the correct module structure and init function name.

Module structure example on mission's side
| modules
|_ YourModule
|__ fn_initYourModule.sqf
|__ fn_function1YourModule.sqf
|__ fn_function2YourModule.sqf

Module structure example on server's side
| modules
|_ YourModule
|__ fn_initYourModule.sqf
|__ fn_serversideFunction1YourModule.sqf

You can check out the example module for more information on structuring

You can also add actions to the interaction menu in any of your scripts, but it should be done in your init file or by calling another function initializing menu actions first in your init.
Interaction menu used is CBA's Fleximenu and this is the menu action array format for adding and some additional scripting to help you get started. Just make sure to append these arrays to the RPF_MenuItems array.
```javascript
_additionalMenuItems = [
  ["Example", {
  [] call Client_fnc_hintExample;
  true
  }, "", "", [], -1, true, (CONDITION)]
];
{
  RPF_MenuItems pushBack _x;
}forEach _additionalMenuItems;
```
Check the example module for a good example!

Now important step is making sure to include a class with your functions that will be added to the mission's Functions.hpp or server's config.cpp.
Proper way of doing the CfgFunctions class:
Basically, your classes correspond to the filenames in your module's folder without the fn_ part and .sqf ending. For the CfgFunctions part in the mission, you will have a different path than the server to your module folder so make sure you notice this when you write the file line of your class.

This would go to mission > Functions.hpp and under class ClientModules which is in CfgFunctions
```C++
class YourModule {
  file = "Functions\modules\YourModule";
  class initYourModule {};
  class function1YourModule {};
  class function2YourModule {};
};
```

This one would go to RPF_server > config.cpp and under class ServerModules which is in CfgFunctions
```C++
class YourModule {
  file = "RPF_server\server\modules\YourModule";
  class initYourModule {};
  class serversideFunction1YourModule {};
};
```

**Comment by superihippo** [comment]
*Module documentation for 0.0.5*

1. General
Modules are additional script packages that might depend on the core modules, but the core never depends on the module. Modules can overwrite settings, add functionality, add interaction menu items/options and much more. Modules can be distributed under different licenses than the core of RPFramework.

The core is a bunch of basic functionality that creates a fairly limited game experience and loads in modules. 

Modules are divided into "Default Modules" and "Additional Modules". Default modules add default functionality to the framework, but can be easily disabled if one wants to change their functionality by an additional module. Difference between core and Default Modules is that core is standalone, but Default Modules depend on core and default modules can of course be disabled and it wont break the core.

**Comment by superihippo** [comment]
2. Installing modules

Installing additional modules should be instructed by the module creator as different modules might have different features but there is a general instruction:
1. Move module's files to mission>Functions>modules and RPF_Server>server>modules
2. Add module's function class into mission>Functions.hpp in the class ClientModules and into RPF_Server>config.cpp in the class ServerModules
3. Open up mission>Functions>core>init>fn_initModules.sqf and RPF_server>server>core>init>fn_initModules.sqf and in the mission's file add your module's init function to the additional modules array in format "ClientModule_fnc_initNameOfYourModule" and in the server's file add your module's init function to the additional modules array in format "ServerModule_fnc_initNameOfYourModule".
4. Your module should be installed properly.

**Comment by superihippo** [comment]
3. Creating a module

Modules add features to the RPFramework. If you are a scripter, you already know how to make modules because they are just bunch of functions to the mission and server.

Modules aren't anything special, but they just make things a bit more modular and easier for the user of the framework.

You can add any functions, sounds and dialogs if you just make sure to instruct the user of your module to add them properly. It is also very important that you follow the correct module structure and init function name.

Module structure example on mission's side
| modules
|_ YourModule
|__ fn_initYourModule.sqf
|__ fn_function1YourModule.sqf
|__ fn_function2YourModule.sqf

Module structure example on server's side
| modules
|_ YourModule
|__ fn_initYourModule.sqf
|__ fn_serversideFunction1YourModule.sqf

You can check out the example module for more information on structuring

You can also add actions to the interaction menu in any of your scripts, but it should be done in your init file or by calling another function initializing menu actions first in your init.

INSERT PROPER FLEXIMENU ARRAY STRUCTURE HERE
See modules>example>fn_initExampleMenu.sqf
_additionalMenuItems = [
            ["Example", {
                [] call Client_fnc_hintExample;
                true
            }, "", "", [], -1, true, (CONDITION)]
];
{
	RPF_MenuItems pushBack _x;
}forEach _additionalMenuItems;

Now important step is making sure to include a class with your functions that will be added into the mission's Functions.hpp or server's config.cpp.
Proper way of doing the CfgFunctions class:
Basically your classes correspond to the filenames in your module's folder without the fn_ part and .sqf ending. For the CfgFunctions part in the mission you will have different path do your module folder so make sure you notice this when you write the file line of your class.

This would go to mission > Functions.hpp and under class ClientModules which is in CfgFunctions
class YourModule {
   file = "Functions\modules\YourModule";
   class initYourModule {};
   class function1YourModule {};
   class function2YourModule {};
};

This one would go to RPF_server > config.cpp and under class ServerModules which is in CfgFunctions
class YourModule {
   file = "RPF_server\server\modules\YourModule";
   class initYourModule {};
   class serversideFunction1YourModule {};
};

**Comment by superihippo** [comment]
4. Benefits of a modular approach

So why should you hop on board with the modular approach to this framework? 
This modular approach isn't a good approach for a normal mission but for a framework it has been proven effective. Think of arma 3 as a framework and mods as modules; modules can be easily enabled and disabled by the user and of course installed easily without messing with the core files of the game. If you got the example you might already think it's a good approach for this framework as proven by arma. In the RPFramework installing modules isn't as easy as in Arma 3, but it gets pretty close. I have seen the attempts of modifying missions like altis life that people have tried to use as a framework and it gets really messy, but this way you keep your modules and original files from the core separated.
